yarn-project/sequencer-client/src/sequencer/public_processor.ts

```diff
@@ -73,65 +91,133 @@ export class PublicProcessor {
+      }
+
+      const vk = getVerificationKeys().publicKernelCircuit;
+      const vkIndex = 0;
+      const vkSiblingPath = MembershipWitness.random(VK_TREE_HEIGHT).siblingPath;
+      const previousKernel = new PreviousKernelData(previousOutput, previousProof, vk, vkIndex, vkSiblingPath);
+      const inputs = new PublicKernelInputs(previousKernel, callData);
+      // TODO: This should be set by the public kernel circuit
+      inputs.previousKernel.publicInputs.end.publicCallCount = new Fr(1n);
+      return this.publicKernel.publicKernelCircuitNonFirstIteration(inputs);
+    } else {
+      const inputs = new PublicKernelInputsNoPreviousKernel(txRequest, callData);
+      return this.publicKernel.publicKernelCircuitNoInput(inputs);
     }
+  }
+
+  protected async getPublicCircuitPublicInputs(result: PublicExecutionResult) {
+    const publicDataTreeInfo = await this.db.getTreeInfo(MerkleTreeId.PUBLIC_DATA_TREE);
+    const historicPublicDataTreeRoot = Fr.fromBuffer(publicDataTreeInfo.root);
+    const callStackPreimages = await this.getPublicCallStackPreimages(result);
+    const wasm = await CircuitsWasm.get();
+    const publicCallStack = callStackPreimages.map(item => computeCallStackItemHash(wasm, item));
 
-    return new PublicCallData(
-      callStackItem,
-      publicCallStackPreimages,
-      publicCircuitProof,
-      portalContractAddress,
-      bytecodeHash,
+    return PublicCircuitPublicInputs.from({
+      args: result.execution.args,
+      callContext: result.execution.callContext,
+      proverAddress: AztecAddress.random(),
+      emittedEvents: padArrayEnd([], Fr.ZERO, EMITTED_EVENTS_LENGTH),
+      newL2ToL1Msgs: padArrayEnd([], Fr.ZERO, NEW_L2_TO_L1_MSGS_LENGTH),
+      returnValues: padArrayEnd(result.returnValues, Fr.ZERO, RETURN_VALUES_LENGTH),
+      stateReads: padArrayEnd(result.stateReads, StateRead.empty(), STATE_READS_LENGTH),
+      stateTransitions: padArrayEnd(result.stateTransitions, StateTransition.empty(), STATE_TRANSITIONS_LENGTH),
+      publicCallStack,
+      historicPublicDataTreeRoot,
+    });
+  }
+
+  protected async getPublicCallStackItem(result: PublicExecutionResult) {
+    return new PublicCallStackItem(
+      result.execution.contractAddress,
+      result.execution.functionData,
+      await this.getPublicCircuitPublicInputs(result),
     );
   }
+
+  protected async getPublicCallStackPreimages(result: PublicExecutionResult) {
+    const nested = result.nestedExecutions;
+    const preimages: PublicCallStackItem[] = await Promise.all(nested.map(n => this.getPublicCallStackItem(n)));
+    if (preimages.length > PUBLIC_CALL_STACK_LENGTH) {
+      throw new Error(`Public call stack size exceeded (max ${PUBLIC_CALL_STACK_LENGTH}, got ${preimages.length})`);
+    }
+
+    const emptyPreimage = PublicCallStackItem.empty();
+    // TODO: Remove the msgSender set once circuits dont validate empty call stack items
```

Comment:
This should already be done @spalladino

Reply:
I'm getting an error if I comment that line out. I'll create an issue to tackle it separately.

```
Error(3020): call_stack_msg_sender[0] = 0x0000000000000000000000000000000000000000000000000000000000000000 expected 0x0000000000000000000000000000000000000000000000000000000000000020; does not reconcile
```

Reply:
https://github.com/AztecProtocol/aztec-packages/issues/494